lib/ansible/modules/cloud/amazon/elasticache_facts.py

```diff
@@ -0,0 +1,319 @@
+      returned: always
+      type: complex
+      sample:
+        Application: web
+        Environment: test
+'''
+
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, get_aws_connection_info
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict, HAS_BOTO3, AWSRetry
+from ansible.module_utils.ec2 import boto3_tag_list_to_ansible_dict
+
+
+try:
+    import botocore
+except ImportError:
+    pass  # caught by imported HAS_BOTO3
+
+
+@AWSRetry.exponential_backoff()
+def describe_cache_clusters_with_backoff(client, cluster_id=None):
+    paginator = client.get_paginator('describe_cache_clusters')
+    params = dict(ShowCacheNodeInfo=True)
+    if cluster_id:
+        params['CacheClusterId'] = cluster_id
+    try:
+        response = paginator.paginate(**params).build_full_result()
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'CacheClusterNotFound':
+            return []
+        else:
+            raise
+    return response['CacheClusters']
+
+
+@AWSRetry.exponential_backoff()
+def get_elasticache_tags_with_backoff(client, cluster_id):
+    return client.list_tags_for_resource(ResourceName=cluster_id)['TagList']
+
+
+def get_aws_account_id(module):
+    try:
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        client = boto3_conn(module, conn_type='client', resource='sts',
+                            region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoCredentialsError as e:
+        module.fail_json(msg="Can't authorize connection - " + str(e))
+
+    try:
+        return client.get_caller_identity()['Account']
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Couldn't obtain AWS account id: %s" % e.message,
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+
+
+def get_elasticache_clusters(client, module, region):
+    try:
+        clusters = describe_cache_clusters_with_backoff(client, cluster_id=module.params.get('name'))
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Couldn't obtain cache cluster info: %s" % e.message,
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+
+    account_id = get_aws_account_id(module)
+    results = []
+    for cluster in clusters:
+
+        cluster = camel_dict_to_snake_dict(cluster)
+        arn = "arn:aws:elasticache:%s:%s:cluster:%s" % (region, account_id, cluster['cache_cluster_id'])
+        try:
+            tags = get_elasticache_tags_with_backoff(client, arn)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg="Couldn't get tags for cluster %s: %s" % (cluster['cache_cluster_id'], e.message),
```

Comment:
Using e.message causes exception handling failure on python3.
I deleted my cluster mid-call to get an example of this:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 286, in get_elasticache_clusters
    tags = get_elasticache_tags_with_backoff(client, arn)
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_modlib.zip/ansible/module_utils/cloud.py", line 154, in retry_func
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 252, in get_elasticache_tags_with_backoff
    return client.list_tags_for_resource(ResourceName=cluster_id)['TagList']
  File "/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py", line 312, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py", line 601, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.CacheClusterNotFoundFault: An error occurred (CacheClusterNotFound) when calling the ListTagsForResource operation: shertel-ec-0001-001 is either not present or not available

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 319, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 315, in main
    module.exit_json(elasticache_clusters=get_elasticache_clusters(client, module, region))
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 288, in get_elasticache_clusters
    module.fail_json(msg="Couldn't get tags for cluster %s: %s" % (cluster['cache_cluster_id'], e.message),
AttributeError: 'CacheClusterNotFoundFault' object has no attribute 'message'
```
Another reason to use AnsibleAWSModule and move toward [smart exception handling](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/core.py#L124-L127).